basics/trey-hunner-zip-and-enumerate.md

```diff
@@ -76,9 +76,110 @@ c 3
+
+```python
+>>> users = ["Tushar", "Aman", "Anurag", "Sohit"]
+>>> uids = ["usr122", "usr123", "usr124", "usr125"]
+>>> user_details = zip(uids, users)
+>>> print(list(user_details))
+[('usr122', 'Tushar'), ('usr123', 'Aman'), ('usr124', 'Anurag'), ('usr125', 'Sohit')]
+>>>
+```
+
+If the lists are of different lengths, some items from the end of the longer list will be ignored.
+
+```python
+>>> users = ["Tushar", "Aman", "Anurag"]
+>>> emails = ["tushar@example.com", "aman@example.com", "anurag@example.com", "sohit@example.com"]
+>>> users_contact = zip(users, emails)
+>>> print(list(users_contact))
+[('Tushar', 'tushar@example.com'), ('Aman', 'aman@example.com'), ('Anurag', 'anurag@example.com')]
+>>>
+```
+
+Here the shortest list is `users`, with length 3, so `zip(users, emails)` only takes the first 3 emails.
+> We do not recommend calling `zip()` with lists of different lengths.
+
+### Using zip with `for` loop
+
+It is very common to `for` loop over a `zip()`, and unpack the returned tuples in the `for` loop.
+This is why we introduced unpacking in the beginning of this page.
+
+```python
+>>> roll_nums = [20, 25, 28]
+>>> students = ["Joe", "Max", "Michel"]
+>>> for roll_num, student in zip(roll_nums, students):
+...     print(f"Roll number of {student} is {roll_num}")
+...
+Roll number of Joe is 20
+Roll number of Max is 25
+Roll number of Michel is 28
+>>>
+```
+
+## enumerate
+
+`enumerate()` is an amazing Built-in function offered by python. When used, gives us the index and the item combined. long story short! `enumerate` comes in handy while working with loops as it gives us access to both the item and it's index at the same time.
+
+```python
+>>> even_nums = [2, 4, 6, 8, 10, 12]
+>>> print(list(enumerate(even_nums)))
+[(0, 2), (1, 4), (2, 6), (3, 8), (4, 10), (5, 12)]
+>>>
+```
+
+### looping with `enumerate()`
+
+```python
+>>> even_nums = [2, 4, 6, 8, 10, 12]
+>>> for index, item in enumerate(even_nums):
+...     print(f"Index of {item} is {index}")
+...
+Index of 2 is 0
+Index of 4 is 1
+Index of 6 is 2
+Index of 8 is 3
+Index of 10 is 4
+Index of 12 is 5
+>>>
+```
+
+### looping without `enumerate()`
+
+let's take a look at how we can do the same thing in a more traditional way.
+
+```python
+>>> even_nums = [2, 4, 6, 8, 10, 12]
+>>> for index in range(0,  len(even_nums)):
+...     print(f"Index of {even_nums[index]} is {index}")
+...
+Index of 2 is 0
+Index of 4 is 1
+Index of 6 is 2
+Index of 8 is 3
+Index of 10 is 4
+Index of 12 is 5
+>>>
+```
+
+* `range(0, len(even_nums))` gives 0,1,2,3,4,5, with the list length 6 excluded
+* because these are the valid indexes of a list of length 6, `even_nums[index]` prints each element of `even_nums`
```

Comment:
```suggestion
Here:
* `range(0, len(even_nums))` gives 0,1,2,3,4,5, with the list length 6 excluded. These are the indexes of our list of length 6.
* `even_nums[index]` prints each element of `even_nums`, because `index` comes from the range of all indexes into that list.
```